File: t/PACSFramework/VB/StorageServerManager/DataTypes/MachineIDs.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Management

Namespace Leadtools.Demos.StorageServer.DataTypes
   Public Class MachineIDs
      Private Shared _ins As MachineIDs = New MachineIDs()

      Public Shared ReadOnly Property Instance() As MachineIDs
         Get
            Return _ins
         End Get
      End Property

      Private _sMACAddress As List(Of String) = Nothing
      Private _sHDSerial As List(Of String) = Nothing
      Private _sCPUId As String = Nothing

      Public ReadOnly Property MACAddress() As List(Of String)
         Get
            If Nothing Is _sMACAddress Then
               _sMACAddress = GetMACAddress()
            End If
            Return _sMACAddress
         End Get
      End Property

      Public ReadOnly Property HDSerial() As List(Of String)
         Get
            If Nothing Is _sHDSerial Then
               _sHDSerial = GetHDSerial(True)
            End If
            Return _sHDSerial
         End Get
      End Property

      Public ReadOnly Property CPUId() As String
         Get
            If Nothing Is _sCPUId Then
               _sCPUId = GetCPUId()
            End If
            Return _sCPUId
         End Get
      End Property

      Private Function GetHDSerial(ByVal fFirstHDOnly As Boolean) As List(Of String)
         Dim HDSerialsList As List(Of String) = New List(Of String)()
         Dim m As MachineIDs = MachineIDs.Instance
         Dim myScope As ManagementScope = New ManagementScope("root\CIMV2")
         Dim q As SelectQuery = New SelectQuery("Win32_LogicalDisk")
         Dim s As ManagementObjectSearcher = New ManagementObjectSearcher(myScope, q)

         For Each disk As ManagementObject In s.Get()
            Try
               Dim dsk As ManagementObject = New ManagementObject(disk.ToString())
               dsk.Get()
               Dim volumeSerial As String = dsk("VolumeSerialNumber").ToString()
               HDSerialsList.Add(volumeSerial)

               If fFirstHDOnly Then
                  Exit Try
               End If
            Catch
            End Try
         Next disk
         Return HDSerialsList
      End Function

      ''' <summary>
      ''' this function is expensive
      ''' </summary>
      ''' <returns></returns>
      Private Function GetCPUId() As String
         Dim cpuInfo As String = String.Empty
         Dim mc As ManagementClass = New ManagementClass("win32_processor")
         Dim moc As ManagementObjectCollection = mc.GetInstances()

         For Each mo As ManagementObject In moc
            If cpuInfo = "" Then
               'Get only the first CPU's ID
               cpuInfo = mo.Properties("processorID").Value.ToString()
               Exit For
            End If
         Next mo
         Return cpuInfo
      End Function

      Private Function GetMACAddress() As List(Of String)
         Dim MACAddressList As List(Of String) = New List(Of String)()
         Dim searcher As ManagementObjectSearcher = New ManagementObjectSearcher("select MACAddress from Win32_NetworkAdapter")

         Try
            For Each share As ManagementObject In searcher.Get()
               For Each PD As PropertyData In share.Properties
                  If Not PD.Value Is Nothing AndAlso PD.Value.ToString() <> "" Then
                     Select Case PD.Value.GetType().ToString()
                        Case "System.String[]"
                           Dim str As String() = CType(PD.Value, String())

                           Dim str2 As String = ""
                           For Each st As String In str
                              str2 &= st & " "
                           Next st

                           MACAddressList.Add(str2)

                        Case "System.UInt16[]"
                           Dim shortData As UShort() = CType(PD.Value, UShort())


                           Dim tstr2 As String = ""
                           For Each st As UShort In shortData
                              tstr2 &= st.ToString() & " "
                           Next st

                           MACAddressList.Add(tstr2)


                        Case Else
                           MACAddressList.Add(PD.Value.ToString())
                     End Select
                  End If
               Next PD
            Next share
         Catch
         End Try

         Return MACAddressList
      End Function

      Public Sub New()
      End Sub
   End Class
End Namespace

File: t/PACSFramework/CS/StorageServerManager/DataTypes/MachineIDs.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Management;

namespace Leadtools.Demos.StorageServer.DataTypes
{
   public class MachineIDs
   {
      static MachineIDs _ins = new MachineIDs();

      public static MachineIDs Instance
      {
         get
         {
            return _ins;
         }
      }

      private List<string> _sMACAddress = null;
      private List<string> _sHDSerial = null;
      private string _sCPUId = null;

      public List<string> MACAddress
      {
         get
         {
            if (null == _sMACAddress)
               _sMACAddress = GetMACAddress();
            return _sMACAddress;
         }
      }

      public List<string> HDSerial
      {
         get
         {
            if (null == _sHDSerial)
               _sHDSerial = GetHDSerial(true);
            return _sHDSerial;
         }
      }

      public string CPUId
      {
         get
         {
            if (null == _sCPUId)
               _sCPUId = GetCPUId();
            return _sCPUId;
         }
      }

      private List<string> GetHDSerial(bool fFirstHDOnly)
      {
         List<string> HDSerialsList = new List<string>();
         MachineIDs m = MachineIDs.Instance;
         ManagementScope myScope =
         new ManagementScope("root\\CIMV2");
         SelectQuery q = new SelectQuery("Win32_LogicalDisk");
         ManagementObjectSearcher s = new ManagementObjectSearcher(myScope, q);

         foreach (ManagementObject disk in s.Get())
         {
            try
            {
               ManagementObject dsk = new ManagementObject(disk.ToString());
               dsk.Get();
               string volumeSerial = dsk["VolumeSerialNumber"].ToString();
               HDSerialsList.Add(volumeSerial);

               if (fFirstHDOnly)
               {
                  break;
               }
            }
            catch
            {
            }
         }
         return HDSerialsList;
      }

      /// <summary>
      /// this function is expensive
      /// </summary>
      /// <returns></returns>
      private string GetCPUId()
      {
         string cpuInfo = string.Empty;
         ManagementClass mc = new ManagementClass("win32_processor");
         ManagementObjectCollection moc = mc.GetInstances();

         foreach (ManagementObject mo in moc)
         {
            if (cpuInfo == "")
            {
               //Get only the first CPU's ID
               cpuInfo = mo.Properties["processorID"].Value.ToString();
               break;
            }
         }
         return cpuInfo;
      }

      private List<string> GetMACAddress()
      {
         List<string> MACAddressList = new List<string>();
         ManagementObjectSearcher searcher = new ManagementObjectSearcher("select MACAddress from Win32_NetworkAdapter");

         try
         {
            foreach (ManagementObject share in searcher.Get())
            {
               foreach (PropertyData PD in share.Properties)
               {
                  if (PD.Value != null && PD.Value.ToString() != "")
                  {
                     switch (PD.Value.GetType().ToString())
                     {
                        case "System.String[]":
                           string[] str = (string[])PD.Value;

                           string str2 = "";
                           foreach (string st in str)
                              str2 += st + " ";

                           MACAddressList.Add(str2);

                           break;
                        case "System.UInt16[]":
                           ushort[] shortData = (ushort[])PD.Value;


                           string tstr2 = "";
                           foreach (ushort st in shortData)
                              tstr2 += st.ToString() + " ";

                           MACAddressList.Add(tstr2);

                           break;

                        default:
                           MACAddressList.Add(PD.Value.ToString());
                           break;
                     }
                  }
               }
            }
         }
         catch
         {
         }

         return MACAddressList;
      }

      public MachineIDs()
      {
      }
   }
}
